### PR TITLE
Add Binary64 option for account data

### LINKED ISSUE #10140: RPC uses base58 for large binary leading to slow response
#### Problem

RPC responses encodes `account.data` with base58:

https://github.com/solana-labs/solana/blob/ad7b113944a62e670a2277503f251ef53b0267a5/client/src/rpc_response.rs#L78

base58 isn't suited for large arbitrary-sized binary blobs. it's only for hash: https://github.com/solana-labs/solana/pull/9490#discussion_r416430038

And it's very easy to attack dos. Also, viewing program account data (which tends to be large by definition) takes long time to retrieve via RPC.

```
$ time curl -s -X POST -H 'Content-Type: application/json' -d '{"id":1,"jsonrpc":"2.0","method":"getAccountInfo","params":["SysvarS1otHistory11111111111111111111111111",{"commitment":"max"}]}' http://tds.solana.com -o - | wc -c
179217

real    0m24.143s
user    0m0.005s
sys     0m0.007s

$ time curl -s -X POST -H 'Content-Type: application/json' -d '{"id":1,"jsonrpc":"2.0","method":"getAccountInfo","params":["SysvarRecentB1ockHashes11111111111111111111",{"commitment":"max"}]}' http://tds.solana.com -o - | wc -c
8389

real    0m0.377s
user    0m0.005s
sys     0m0.005s

$ time curl -s -X POST -H 'Content-Type: application/json' -d '{"id":1,"jsonrpc":"2.0","method":"getAccountInfo","params":["SysvarC1ock11111111111111111111111111111111",{"commitment":"max"}]}' http://tds.solana.com -o - | wc -c
239

real    0m0.304s
user    0m0.002s
sys     0m0.010s

```

As predicted theoretically, this is quadratic time increase to the size of account.data:

```
$ pry
[9] pry(main)> 179217.0 / 8389
=> 21.363332935987604
[11] pry(main)> (377 - 304) * 21.363332935987604 ** 2
=> 33316.615571771246    # =~ 24.143s
```

#### Proposed Solution

Use base64? But compat... ;)

FYI: @CriesofCarrots 

Discovered via #9490.

### PR DESCRIPTION
**Pull request by sakridge**
#### Problem

bs58 encoding of large data is slow

#### Summary of Changes

Add base64 option and limit bs58 requests to account data less that 128 bytes.

Fixes #10140

### CODE REVIEW

File: core/src/rpc.rs [hunk 1]
```diff
@@ -243,15 +243,19 @@ impl JsonRpcRequestProcessor {
...
         new_response(
             &bank,
             bank.get_account(pubkey).and_then(|account| {
                 if account.owner == spl_token_id_v1_0() && encoding == UiAccountEncoding::JsonParsed
                 {
                     get_parsed_token_account(bank.clone(), account)
                 } else {
-                    Some(UiAccount::encode(account, encoding, None))
+                    if encoding == UiAccountEncoding::Binary && account.data.len() > 128 {
+                        None
```
Comment (mvines) [inline core/src/rpc.rs:255]: I think we should return an error back over RPC here, so the user knows they got rejected

**Comment by mvines** [comment]
We'll need an update to `docs/src/apps/jsonrpc-api.md` here eventually.  Also let's convert `solana account`/`solana vote-account`/`solana stake-account` over to use base64 (there may be a couple other commands in the cli that will now die with the 128 byte limit too)

**Comment by mvines** [comment]
Allowing the user to optionally supply an "offset"/"length" seems pretty nice too, for when if they know they only need a slice of a large account

File: core/src/rpc.rs (hunk 1 above)
Comment (sakridge) [reply core/src/rpc.rs]: added

File: docs/src/apps/jsonrpc-api.md [hunk 2]
```diff
@@ -156,7 +156,7 @@ Returns all information associated with the account of provided Pubkey
 - `<string>` - Pubkey of account to query, as base-58 encoded string
 - `<object>` - (optional) Configuration object containing the following optional fields:
   - (optional) [Commitment](jsonrpc-api.md#configuring-state-commitment)
-  - (optional) `encoding: <string>` - encoding for Account data, either "binary" or jsonParsed". If parameter not provided, the default encoding is binary.
+  - (optional) `encoding: <string>` - encoding for Account data, either "binary", "binary64", or jsonParsed". If parameter not provided, the default encoding is "binary". "binary" is bs58 encoded and limited to data queries of less than 128 bytes. Binary64 will return base64 encoded data.
```
Comment (CriesofCarrots) [inline docs/src/apps/jsonrpc-api.md:159]:
```suggestion
  - (optional) `encoding: <string>` - encoding for Account data, either "binary", "binary64", or jsonParsed". If parameter not provided, the default encoding is "binary". "binary" is bs58 encoded and limited to Account data of less than 128 bytes. "binary64" will return base64 encoded data for Account data of any size.
```

Comment (sakridge) [reply docs/src/apps/jsonrpc-api.md]: I took the suggestion.

**Comment by sakridge** [comment]
> 
> 
> Approach looks fine to me. I haven't examined to see why the cli/deploy test is unhappy yet.
> I like the idea of an offset/length specification, and could help with that in a follow-up

rpc_client was trying to decode as Binary instead of Binary64. After I fixed that it seems to be ok, although my fix is a bit hacky.
https://github.com/solana-labs/solana/pull/11474/commits/773f002980b24702ba3d79181055ad16b6b4de6e

**Comment by CriesofCarrots** [comment]
> rpc_client was trying to decode as Binary instead of Binary64. After I fixed that it seems to be ok, although my fix is a bit hacky.

Ah, serde_json must go with the first enum variant that matches type 😕 
I'm fine with your fix; the less-hacky ways I've thought to do this are a lot less concise.

Looks like you just need to update this line to Binary64 to fix the CI failure: https://github.com/solana-labs/solana/blob/a52702146bbba105ad7747e190ceeb08d48059cc/cli/src/offline/blockhash_query.rs#L353

**Comment by mvines** [comment]
> I like the idea of an offset/length specification, and could help with that in a follow-up

@CriesofCarrots - sure, can you queue this up for next week.   We now do have a real use case where it would be more efficient for the client (and the node) if they were able to only request partial account data

File: docs/src/apps/jsonrpc-api.md [hunk 3]
```diff
@@ -156,7 +156,7 @@ Returns all information associated with the account of provided Pubkey
 - `<string>` - Pubkey of account to query, as base-58 encoded string
 - `<object>` - (optional) Configuration object containing the following optional fields:
   - (optional) [Commitment](jsonrpc-api.md#configuring-state-commitment)
-  - (optional) `encoding: <string>` - encoding for Account data, either "binary" or jsonParsed". If parameter not provided, the default encoding is binary.
+  - (optional) `encoding: <string>` - encoding for Account data, either "binary", "binary64", or jsonParsed". If parameter not provided, the default encoding is "binary". "binary" is bs58 encoded and limited to Account data of less than 128 bytes. "binary64" will return base64 encoded data for Account data of any size.
```
Comment (mvines) [inline docs/src/apps/jsonrpc-api.md:159]:
```suggestion
  - (optional) `encoding: <string>` - encoding for Account data, either "binary", "binary64", or jsonParsed". If parameter not provided, the default encoding is "binary". "binary" is base-58 encoded and limited to Account data of less than 128 bytes. "binary64" will return base64 encoded data for Account data of any size.
```

File: core/src/rpc.rs [hunk 4]
```diff
@@ -240,21 +240,27 @@ impl JsonRpcRequestProcessor {
...
-                    Some(UiAccount::encode(account, encoding, None))
-                }
-            }),
-        )
+        let mut response = None;
+        if let Some(account) = bank.get_account(pubkey) {
+            if account.owner == spl_token_id_v1_0() && encoding == UiAccountEncoding::JsonParsed {
+                response = get_parsed_token_account(bank.clone(), account);
+            } else if encoding == UiAccountEncoding::Binary && account.data.len() > 128 {
+                let message = "Encoded binary (bs58) data should be less than 128 bytes, please use Binary64 encoding.".to_string();
```
Comment (mvines) [inline core/src/rpc.rs:252]:
```suggestion
                let message = "Encoded binary (base 58) data should be less than 128 bytes, please use Binary64 encoding.".to_string();
```

File: docs/src/apps/jsonrpc-api.md (hunk 3 above)
Comment (sakridge) [reply docs/src/apps/jsonrpc-api.md]: changed

File: core/src/rpc.rs (hunk 4 above)
Comment (sakridge) [reply core/src/rpc.rs]: changed

**Comment by sakridge** [comment]
```
test bench_encode_base58 ... bench:  20,546,592 ns/iter (+/- 65,220)
test bench_encode_base64 ... bench:   6,253,972 ns/iter (+/- 107,235)
```
base64 does seem quite a bit faster.

**Comment by mvines** [comment]
Definitely faster. I did a `./run.sh` with/without this PR and timed `solana accounts TokenSVp5gheXUvJ6jGWGeCsgPKgnE3YgdGKRVCMY9o` -- 8 seconds for base58, 3.5 seconds for base64. 

I wonder if we could do even better with just hex encoding (as an alternative for when increase in encoding size doesn't matter, like small partial account reads) , would that be quick to bench @sakridge?

**Comment by sakridge** [comment]
> Definitely faster. I did a `./run.sh` with/without this PR and timed `solana accounts TokenSVp5gheXUvJ6jGWGeCsgPKgnE3YgdGKRVCMY9o` -- 8 seconds for base58, 3.5 seconds for base64.
> 
> I wonder if we could do even better with just hex encoding (as an alternative for when increase in encoding size doesn't matter, like small partial account reads) , would that be quick to bench @sakridge?

Hex seems worse:
```
test bench_encode_base58 ... bench:  73,587,849 ns/iter (+/- 531,971)
test bench_encode_base64 ... bench:       3,518 ns/iter (+/- 31)
test bench_encode_hex    ... bench:      26,636 ns/iter (+/- 84)
```
That's 8k of random data which with base58 is all I can stand to wait for. It really falls apart with random data.

**Comment by mvines** [comment]
Hex is worse 🤯.

**Comment by sakridge** [comment]
> Hex is worse 🤯.

yep @mvines any blockers for merging this?

**Pull request merged**